### PR TITLE
Develop

### PR DESCRIPTION
**Pull request by lulys96**
New software architecture, memspecs altered to have JEDEC compliance, DDR3 and DDR4 supported.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -3,6 +3,35 @@ cmake_minimum_required(VERSION 3.10)
...
 
-include_directories(src/)
+include_directories(src/
+                    src/dramtypes
+                    src/memspec
+                    src/cli)
 
-add_library(DRAMPower src/libdrampower/LibDRAMPower.cc)
+add_library(DRAMPower src/cli/drampower.cc
+            src/cli/CliHandler.cc
```
Comment (myzinsky) [inline CMakeLists.txt]: CLI Handler should not be part of the DRAMPower library.

File: CMakeLists.txt [hunk 2]
```diff
@@ -3,6 +3,35 @@ cmake_minimum_required(VERSION 3.10)
...
 project(DRAMPower)
 
-include_directories(src/)
+include_directories(src/
+                    src/dramtypes
+                    src/memspec
+                    src/cli)
 
-add_library(DRAMPower src/libdrampower/LibDRAMPower.cc)
+add_library(DRAMPower src/cli/drampower.cc
```
Comment (myzinsky) [inline CMakeLists.txt]: - This drampower.cc should not be part of the library, a library does not have a ```main()```
- Naming should be drampower_cli.cc
- Call the library **DRAMPowerLibrary**

File: CMakeLists.txt [hunk 3]
```diff
@@ -3,6 +3,35 @@ cmake_minimum_required(VERSION 3.10)
...
+            src/memspec/MemSpec.cc
+            src/memspec/MemSpecDDR3.cc
+            src/memspec/MemSpecDDR4.cc
+            src/memspec/MemSpecWideIO.cc
+            src/MemCommand.cc
+            src/counters/Counters.cc
+            src/counters/CountersDDR3.cc
+            src/counters/CountersDDR4.cc
+            src/counters/CountersWideIO.cc
+            src/TraceParser.cc
```
Comment (myzinsky) [inline CMakeLists.txt]: Should this not also be part of the CLI?

File: .gitignore [hunk 4]
```diff
@@ -16,3 +16,6 @@ cscope.*
 commands.trace
 coverage_report
 .DS_Store
+CMakeFiles/*
+CmakeLists.txt.user
```
Comment (myzinsky) [inline .gitignore]: It should be CMakeLists.txt.user with capital M

File: CMakeLists.txt.user [hunk 5]
```diff
@@ -0,0 +1,384 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (myzinsky) [inline CMakeLists.txt.user]: This file should not be in GIT!!

File: DebugFile.txt [hunk 6]
```diff
@@ -0,0 +1,24 @@
+ 	: Parameter ipp2n not found: parsed with zero.
```
Comment (myzinsky) [inline DebugFile.txt]: delete

File: Makefile [hunk 7]
```diff
@@ -1,149 +1,678 @@
-#/*
```
Comment (myzinsky) [inline Makefile:1]: Please delete this file

File: README.md [hunk 8]
```diff
@@ -12,6 +12,7 @@ The master branch of the repository should be regarded as the bleeding-edge vers
 
 Clone the repository, or download the zip file of the release you would like to use. The source code is available in src folder. [drampower.cc](src/cli/drampower.cc) file gives the user interface, where the user can specify the memory to be employed and the commandtrace to be analyzed. To build, use:
 ```bash
+cmake .
```
Comment (myzinsky) [inline README.md]: Here we could create a build folder:

```bash
mkdir build
cd build
cmake ../
make
```

And then put the build in the gitignore file :)

File: README.md [hunk 9]
```diff
@@ -31,10 +32,9 @@ The tool was verified on Ubuntu >= 14.04. For building a gcc or clang is require
 If the command-level interface is being used, a command trace can be logged in a file.
 An example is given in ```traces/commands.trace```
 
-The format it uses is: ```<timestamp>,<command>,<bank>```.
-For example, "500,ACT,2", where ACT is the command and 2 is the bank. Timestamp is in clock cycles (cc), the list of supported commands is
-mentioned in [MemCommand.h](src/MemCommand.h) and the bank is the target bank number. For non-bank-specific commands, bank can be set to 0. Rank need not be
-specified. The timing correctness of the trace is not verified by the tool and is assumed to be accurate. However, warning messages are provided, to identify if the memory or bank state is inconsistent in the trace. A sample command trace is provided in the traces/ folder.
+The format it uses is: ```<timestamp>,<command>,<rank>, <bank>```.
```
Comment (myzinsky) [inline README.md]: put spaces after comma.

File: README.md [hunk 10]
```diff
@@ -45,40 +45,125 @@ This feature is obsolete and not supported any more. One can check out [commit](
 
 To list out the CLI options supported by DRAMPower, type
 ```bash
-./drampower -h
+./DRAMPowerCli -h
```
Comment (myzinsky) [inline README.md]: This binary should be called **DRAMPower**

File: README.md [hunk 11]
```diff
@@ -45,40 +45,125 @@ This feature is obsolete and not supported any more. One can check out [commit](
...
+It is possible to include bankwise parameters: ρ - ACT Standby bankwise power offset factor, σ - Self-Refresh bankwise power offset factor. If not included both will be by default defined as 100.
+
+For DDR3 and DDR4 it is also possible to enable partial array self refresh and define the PASR mode. Example:
+
+```
+	"bankwisespec": {
```
Comment (myzinsky) [inline README.md]: take care if indention here...

File: README.md [hunk 12]
```diff
@@ -45,40 +45,125 @@ This feature is obsolete and not supported any more. One can check out [commit](
...
+		"factRho":100,
+		"factSigma":100,
+		"hasPASR": true,
+		"pasrMode": 1
+	}
+```
+
+For DDR4 it is possible to specify the refresh mode as "1X", "2X" or "4X". If not specified the refresh mode will be by default defined as "1X". Example:
+```
+"RefreshMode": "1X",
```
Comment (myzinsky) [inline README.md:94]: This should be 1, 2, 4 not 1X, 2X, 4X

File: README.md [hunk 13]
```diff
@@ -45,40 +45,125 @@ This feature is obsolete and not supported any more. One can check out [commit](
...
+            "XS": 64,
+            "XSDLL": 512,
+            "ACTPDEN": 1,
+            "PRPDEN": 1,
+            "REFPDEN": 1,
+            "RTRS": 1,
+            "clkMhz": 533
+        },
+	
+       	"bankwisespec": {
```
Comment (myzinsky) [inline README.md:159]: take care on indention!

File: README.md [hunk 14]
```diff
@@ -100,101 +185,21 @@ To include these JSONs in your simulations, simply use them as the target memory
...
 
 An example of using this tool is provided below. Clone the repository and make.
-```
-make -j4
-```
+
 Once the build is successful, run with the command trace for combined bank simulation:
 ```
-./drampower -m memspecs/MICRON_1Gb_DDR3-1066_8bit_G.json -c traces/commands.trace -r
+./DRAMPowerCli -m memspecs/MICRON_1Gb_DDR3-1066_8bit_G.json -c traces/commands.trace -r
```
Comment (myzinsky) [inline README.md]: ./DRAMPower is better than DRAMPowerCli

File: README.md [hunk 15]
```diff
@@ -100,101 +185,21 @@ To include these JSONs in your simulations, simply use them as the target memory
...
 ```
 The output should be something like this:
```
Comment (myzinsky) [inline README.md]: I think we can remove this example output. It makes the readme just blowing up...

File: cmake_install.cmake [hunk 16]
```diff
@@ -0,0 +1,49 @@
+# Install script for directory: /home/luiza/Hiwi/DRAMPower
```
Comment (myzinsky) [inline cmake_install.cmake]: please remove this file

File: src/MemCommand.cc [hunk 17]
```diff
@@ -32,125 +33,72 @@
...
 using namespace DRAMPower;
 using namespace std;
 
 
-MemCommand::MemCommand(MemCommand::cmds type,
-                       unsigned bank, int64_t timestamp) :
-  type(type),
-  bank(bank),
-  timestamp(timestamp)
+MemCommand::MemCommand( int64_t timestamp, MemCommand::cmds type,
```
Comment (myzinsky) [inline src/MemCommand.cc]: remove initial whitespace

File: src/MemCommand.h [hunk 18]
```diff
@@ -69,122 +69,122 @@ class MemCommand {
...
-      if (typeStrings[typeId] == name) {
-        cmds commandType = static_cast<cmds>(typeId);
-        return commandType;
-      }
+
+    static const unsigned int nCommands = 20;
+
+    static std::string* getCommandTypeStrings()
+    {
+        static std::string type_map[nCommands] = { "ACT",
```
Comment (myzinsky) [inline src/MemCommand.h]: maybe this map shal be defined near to the enum? Such that we direcetly see what to update if a new command is inserted.

File: src/MemCommand.h [hunk 19]
```diff
@@ -69,122 +69,122 @@ class MemCommand {
...
+                                                   "SREN",
+                                                   "SREX",
+                                                   "NOP",
+                                                   "UNINITIALIZED" };
+
+        return type_map;
+    }
+
+    // To identify command type from name
+    static cmds getTypeFromName(const std::string& name)
```
Comment (myzinsky) [inline src/MemCommand.h]: Please just keep prototypes in header file and bring the actual implementation to CC files.

File: src/TraceParser.cc [hunk 20]
```diff
@@ -1,101 +1,106 @@
-/*
```
Comment (myzinsky) [inline src/TraceParser.cc:1]: Can you keep here the
```c++
/*
*
*
*/
```
styple for the license comment?

File: src/TraceParser.h [hunk 21]
```diff
@@ -43,25 +45,26 @@
...
+
 
 
 class TraceParser {
- public:
-  TraceParser(const DRAMPower::MemorySpecification& memSpec);
-  // list of parsed commands
-  std::vector<DRAMPower::MemCommand> cmd_list;
+public:
+    TraceParser(){}
```
Comment (myzinsky) [inline src/TraceParser.h]: why does the trace parser does not have any inputs?

File: src/cli/CliHandler.cc [hunk 22]
```diff
@@ -1,8 +1,8 @@
 /*
- * Copyright (c) 2012-2014, TU Delft
- * Copyright (c) 2012-2014, TU Eindhoven
- * Copyright (c) 2012-2014, TU Kaiserslautern
- * Copyright (c) 2012-2019, Fraunhofer IESE
+ * Copyright (c) 2012-2020, TU Delft
```
Comment (myzinsky) [inline src/cli/CliHandler.cc]: dont change dates of TU Delft and TU Einfhoven, just TUK and IESE :-)

File: src/cli/CliHandler.cc [hunk 23]
```diff
@@ -82,149 +83,148 @@ void CliHandler::logo()
...
-  return pasr_mode;
+void CliHandler::loadMemSpec(const std::string &memspecUri)
+{
+    json doc = traceparser.parseJSON(memspecUri);
+    json jMemSpec = doc["memspec"];
+
+    std::string memoryType = jMemSpec["memoryType"];
+
+    if (memoryType == "DDR3") {
+        MemSpecDDR3 memSpecDDR3(jMemSpec,(get_writeToConsole() | get_writeToFile()),
```
Comment (myzinsky) [inline src/cli/CliHandler.cc]: - Why are get_writeToConsole etc relevant for the memspec. This information should go to the library instead.
- We shoud think on the following: 
    - Lets keep the memspec as in DRAMSys as a pure data class
    - Introduce a configuration loader that parses the memspec json and fills the memspec data objects (code of DRAMSys could be reused here)

File: src/cli/CliHandler.cc [hunk 24]
```diff
@@ -82,149 +83,148 @@ void CliHandler::logo()
...
+        MemSpecDDR4 memSpecDDR4(jMemSpec,(get_writeToConsole() | get_writeToFile()),
+                                get_writeToConsole(),get_writeToFile(),"DebugFile");
+
+        dramPower = new DRAMPowerDDR4(memSpecDDR4,get_io_term_active());
+    }
+    //    else if (memoryType == "LPDDR4")
+    //        memSpec = new MemSpecLPDDR4(jMemSpec);
+    else if (memoryType == "WIDEIO_SDR") {
+
+        MemSpecWideIO memSpecWideIO(jMemSpec,(get_writeToConsole() | get_writeToFile()),
```
Comment (myzinsky) [inline src/cli/CliHandler.cc]: Just a side note, i would format it like this:

```c++
        MemSpecWideIO memSpecWideIO(jMemSpec,
            (get_writeToConsole() | get_writeToFile()),
            get_writeToConsole(),
            get_writeToFile(),
            "DebugFile");
```

File: src/cli/CliHandler.cc [hunk 25]
```diff
@@ -82,149 +83,148 @@ void CliHandler::logo()
...
+                                   get_writeToConsole(), get_writeToFile(),"DebugFile");
+
+        dramPower = new DRAMPowerWideIO(memSpecWideIO, get_io_term_active());    }
+    //    else if (memoryType == "WIDEIO2")
```
Comment (myzinsky) [inline src/cli/CliHandler.cc:206]: This comment shows future work? :-)

File: src/cli/CliHandler.h [hunk 26]
```diff
@@ -1,8 +1,8 @@
 /*
- * Copyright (c) 2012-2014, TU Delft
```
Comment (myzinsky) [inline src/cli/CliHandler.h:2]: same as commented below

File: src/common/version.h [hunk 27]
```diff
@@ -1,8 +1,8 @@
 /*
- * Copyright (c) 2012-2014, TU Delft
```
Comment (myzinsky) [inline src/common/version.h:2]: same

File: src/counters/CountersDDR3.h [hunk 28]
```diff
@@ -0,0 +1,92 @@
...
+  CountersDDR3(MemSpecDDR3& memspec);
+  MemSpecDDR3 memSpec;
+  void getCommands(std::vector<MemCommand>& list,
+                             bool lastupdate,
+                             int64_t timestamp);
+
+
+  // Handlers for commands that are getting processed
+
+  void handleRef(    unsigned bank, int64_t timestamp);
```
Comment (myzinsky) [inline src/counters/CountersDDR3.h]: we think that all overwritten functions should have a virtual and override specifier.

File: src/dramtypes/DRAMPowerDDR3.cc [hunk 29]
```diff
@@ -0,0 +1,469 @@
...
+    // for each command identify timestamp, type and bank
+    for (auto cmd : cmdList) {
+        // For command type
+        int type = cmd.getType();
+        // For command bank
+        unsigned bank = cmd.getBank();
+        // Command Issue timestamp in clock cycles (cc)
+        int64_t timestamp = cmd.getTimeInt64();
+
+        if (bank < memSpec.numberOfBanks) {
```
Comment (myzinsky) [inline src/dramtypes/DRAMPowerDDR3.cc]: can you turn this in a switch case?

File: src/dramtypes/DRAMPowerDDR4.cc [hunk 30]
```diff
@@ -0,0 +1,492 @@
...
+    pup_pre_energy_banks.assign(static_cast<size_t>(nbrofBanks), 0.0);
+    total_energy_banks.assign(static_cast<size_t>(nbrofBanks), 0.0);
+    window_energy_banks.assign(static_cast<size_t>(nbrofBanks), 0.0);
+
+    read_io_energy      = 0.0;
+    write_term_energy   = 0.0;
+    io_term_energy      = 0.0;
+}
+
+void DRAMPowerDDR4::bankEnergyCalc(unsigned vdd)
```
Comment (myzinsky) [inline src/dramtypes/DRAMPowerDDR4.cc]: please use directly the right mps as input, such thatt the calling function can pass this as argument: memSpec.memPowerSpec[vdd];

File: src/dramtypes/DRAMPowerIF.cc [hunk 31]
```diff
@@ -1,8 +1,8 @@
 /*
```
Comment (myzinsky) [inline src/dramtypes/DRAMPowerIF.cc]: I think an interface should not have any non-virtual methods, or?

File: valgrind-out.txt [hunk 32]
```diff
@@ -0,0 +1,1130 @@
+==11615== Memcheck, a memory error detector
```
Comment (myzinsky) [inline valgrind-out.txt:1]: Please do not push stuff like this...

**Review by myzinsky — CHANGES_REQUESTED** [review]
We just did a structural review and have not yet started with checking equations etc.

File: CMakeLists.txt (hunk 3 above)
Comment (lulys96) [reply CMakeLists.txt]: I include the library in the CLI

**Pull request closed without merging**